Lib/pty.py

```diff
@@ -121,12 +121,6 @@ def fork():
     # Parent and child process.
     return pid, master_fd
 
-def _writen(fd, data):
-    """Write all the data to a descriptor."""
-    while data:
-        n = os.write(fd, data)
-        data = data[n:]
-
 def _read(fd):
     """Default read function."""
     return os.read(fd, 1024)
```

Comment:
Would it be okay to replace `high_waterlevel` and this `1024` above here with a `PIPE_BUF` global set to 4096 so it's clear where it comes from? This also makes the deadlock in line 164 (`master_read`) more apparent.

Reply:
`_copy` accept caller provided `master_read` (may be something other than `pty._read`). It does not support "read at most n bytes" by design. The actual buffering here may exceed 4096. `high_waterlevel` serves as a threshold determines whether we should continue reading and appending into our buffer, it can not limit maximum buffering. A name like `PIPE_BUF` is not appropriate per my understanding.

I do want to increase the read size from `1024` to `4096`, but I'm afraid of unexpected changes to application behavior. As long as we didn't pass lots of data via pty, `1024` should serve us well. So I didn't change it.